File: experiments-local/experiment-15/runs/agent=omp_language=go_model=openrouter/minimax/minimax-m3_tooling=none/rep2/internal/book/book.go
// Package book defines the Book domain type used by the API and store layers.
package book

import (
	"errors"
	"fmt"
	"strings"
)

// Book is the canonical representation of a book in the collection.
//
// The JSON tags are the wire format. The store layer maps these columns to a
// SQLite row; the API layer maps these to JSON responses.
type Book struct {
	ID     int64  `json:"id"`
	Title  string `json:"title"`
	Author string `json:"author"`
	Year   int    `json:"year"`
	ISBN   string `json:"isbn"`
}

// Input is the subset of Book fields a client may supply on create/update.
// ID is assigned by the store and never accepted from the wire.
type Input struct {
	Title  string `json:"title"`
	Author string `json:"author"`
	Year   int    `json:"year"`
	ISBN   string `json:"isbn"`
}

// ErrValidation is returned for caller-fixable input problems. The API layer
// translates it into HTTP 400. The wrapped message is human-readable on a
// single line — no embedded newlines, so it can be rendered in a JSON
// {"error": "..."} payload without surprising clients.
var ErrValidation = errors.New("validation failed")

// Validate enforces the API contract: title and author are required (after
// trimming whitespace); year must be non-negative if supplied. ISBN is
// optional. Returns ErrValidation wrapping a descriptive message.
func (in Input) Validate() error {
	title := strings.TrimSpace(in.Title)
	author := strings.TrimSpace(in.Author)
	switch {
	case title == "":
		return joinErr("title is required")
	case author == "":
		return joinErr("author is required")
	case in.Year < 0:
		return joinErr("year must be non-negative")
	}
	return nil
}

// Normalize returns a copy of the input with surrounding whitespace stripped
// from the string fields. Callers should use the result when persisting.
func (in Input) Normalize() Input {
	return Input{
		Title:  strings.TrimSpace(in.Title),
		Author: strings.TrimSpace(in.Author),
		Year:   in.Year,
		ISBN:   strings.TrimSpace(in.ISBN),
	}
}

func joinErr(msg string) error {
	// Use a single-line wrapped error so the API's JSON error field renders
	// cleanly. errors.Join would otherwise embed a newline between the
	// sentinel and the cause, leaking Go's default error formatting.
	return fmt.Errorf("%w: %s", ErrValidation, msg)
}

File: experiments-local/experiment-15/runs/agent=omp_language=typescript_model=openrouter/tencent/hy3-preview_tooling=none/rep1/src/models/Book.ts
export interface Book {
  id: number;
  title: string;
  author: string;
  year?: number;
  isbn?: string;
  created_at: string;
  updated_at: string;
}

export interface CreateBookInput {
  title: string;
  author: string;
  year?: number;
  isbn?: string;
}

export interface UpdateBookInput {
  title?: string;
  author?: string;
  year?: number;
  isbn?: string;
}

export interface ValidationError {
  field: string;
  message: string;
}

export function validateBookInput(data: Partial<CreateBookInput>): ValidationError[] {
  const errors: ValidationError[] = [];

  if (!data.title || data.title.trim().length === 0) {
    errors.push({ field: 'title', message: 'Title is required' });
  }

  if (!data.author || data.author.trim().length === 0) {
    errors.push({ field: 'author', message: 'Author is required' });
  }

  if (data.year !== undefined && (typeof data.year !== 'number' || data.year < 0 || data.year > new Date().getFullYear())) {
    errors.push({ field: 'year', message: 'Year must be a valid number between 0 and current year' });
  }

  if (data.isbn !== undefined && data.isbn && !/^(?=(?:\D*\d){10}(?:(?:\D*\d){3})?$)[\d-]+$/.test(data.isbn)) {
    errors.push({ field: 'isbn', message: 'ISBN must be a valid format' });
  }

  return errors;
}
